lib/NeuraDialect/Mapping/HeuristicMapping/HeuristicMapping.cpp

```diff
@@ -3,42 +3,295 @@
+  // mapped operations as the score.
+  for (int i = 0; i < candidates_to_try; i++) {
+    // Gets the mapping results for each candidate location.
+    NoProducerOpCandidate result = evaluateNoProducerOpCandidate(
+        current_op, current_op_index, candidate_locs[i], i + 1,
+        candidates_to_try, architecture, mapping_state, initial_state);
+
+    if (result.mapped_ops_count > 0) {
+      solutions.push_back(result);
+
+      // Outputs the solution quality information.
+      if (result.fully_mapped) {
+        llvm::outs()
+            << "[GlobalExploration] Found complete mapping solution with "
+            << "mapped=" << result.mapped_ops_count << "/"
+            << getMaterializedOpsWithLevels().size()
+            << ", estimated_II=" << result.current_ii
+            << ", score=" << result.getQualityScore() << "\n";
+      } else {
+        llvm::outs() << "[GlobalExploration] Solution quality: "
+                     << "mapped=" << result.mapped_ops_count << "/"
+                     << getMaterializedOpsWithLevels().size()
+                     << ", fully_mapped="
+                     << (result.fully_mapped ? "yes" : "no")
+                     << ", estimated_II=" << result.current_ii
+                     << ", score=" << result.getQualityScore() << "\n";
+      }
+    }
+  }
+
+  // If we have no solutions, we need to backtrack.
+  if (solutions.empty()) {
+    llvm::outs() << "[GlobalExploration] No feasible solutions found, "
+                    "backtracking\n";
+
+    snapshots.pop_back();
+    candidate_history.pop_back();
+    operation_index_history.pop_back();
+
+    if (snapshots.empty()) {
+      llvm::outs() << "[HeuristicMapping] No more snapshots to restore, "
+                      "mapping failed.\n";
+      return false;
+    }
+    snapshots.back().restore(mapping_state);
+    candidate_history.back()++;
+    return false;
+  }
+
+  // Selects the best solution for mapping this no-producer op based on the
+  // quality score.
+  NoProducerOpCandidate best_solution = *std::max_element(
+      solutions.begin(), solutions.end(),
+      [](const NoProducerOpCandidate &a, const NoProducerOpCandidate &b) {
+        return a.getQualityScore() < b.getQualityScore();
+      });
+
+  best_solution.state.restore(mapping_state);
+
+  llvm::outs() << "[GlobalExploration] Selected best solution with "
+               << "score=" << best_solution.getQualityScore()
+               << ", mapped=" << best_solution.mapped_ops_count
+               << ", fully_mapped="
+               << (best_solution.fully_mapped ? "yes" : "no")
+               << ", estimated_II=" << best_solution.current_ii << "\n";
+
+  // If we have a fully mapped solution, we can finalize the mapping.
+  if (best_solution.fully_mapped) {
+    llvm::outs()
+        << "[HeuristicMapping] Found complete mapping solution through "
+        << "global exploration, finalizing...\n";
+    return true;
```

Comment:
How do we revert the mapping of the mapped ops (after the no-produce op) during trial?